linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,350 @@
+```
+
+`threadpool_close` closes the threadpool.
+
+If `threadpool` is `NULL`, `threadpool_close` shall fail and return.
+
+Otherwise, `threadpool_close` shall join all threads in the `threadpool`.
+
+`threadpool_close` shall call `sm_close_begin`.
+
+`threadpool_close` shall call `sm_close_end`.
+
+### threadpool_schedule_work
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work, THREADPOOL_HANDLE, threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+```
+
+`threadpool_schedule_work` schedule the threadpool to work.
+
+If `threadpool` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+If `work_function` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall get the threadpool state by calling `interlocked_add`.
+
+`threadpool_schedule_work` shall call `sm_exec_begin`.
+
+If `sm_exec_begin` returns `SM_EXEC_REFUSED`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall acquire the SRW lock in shared mode by calling `srw_lock_acquire_shared`.
+
+`threadpool_schedule_work` shall increment the `insert_pos`.
+
+If task state is `TASK_NOT_USED`, `threadpool_schedule_work` shall set the current task state to `TASK_INITIALIZING`.
+
+Otherwise, `threadpool_schedule_work` shall release the shared SRW lock by calling `srw_lock_release_shared` and increase `task_array` capacity by calling `reallocate_threadpool_array`.
+
+  - `reallocate_threadpool_array` shall acquire the SRW lock in exclusive mode by calling `srw_lock_acquire_exclusive`.
+
+  - `reallocate_threadpool_array` shall get the current size of task array by calling `interlocked_add`.
+
+  - If there is any overflow computing the new size, `reallocate_threadpool_array` shall fail and return a non-zero value .
+
+  - Otherwise, `reallocate_threadpool_array` shall double the current task array size.
+
+  - `reallocate_threadpool_array` shall realloc the memory used for the array items and on success return a non-`NULL` handle to it.
+
+  - If any error occurs, `reallocate_threadpool_array` shall fail and return a non-zero value.
+
+  - `reallocate_threadpool_array` shall initialize every task item in the new task array with `task_func` and `task_param` set to `NULL` and `task_state` set to `TASK_NOT_USED`.
+
+  - `reallocate_threadpool_array` shall remove any gap in the task array.
+
+  - `reallocate_threadpool_array` shall reset the `consume_idx` and `insert_idx` to 0 after resize the task array.
+
+  - `reallocate_threadpool_array` shall release the SRW lock by calling `srw_lock_release_exclusive`.
+
+  - `reallocate_threadpool_array` shall return zero on success.
+
+If reallocating the task array fails, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall obtain task information in next available task array index and return zero on success.
```

Comment:
>readpool_schedule_work` shall obtain task information in next available task array index and return zero on success. [](http://example.com/codeflow?start=3&length=116)

The part with shall obtain task information in next available task array is confusing. What does this mean?

Reply:
trying to say get the work function and function context and then store them into the task array here